tests/nvmeof/test_ceph_nvmeof_loadbalancing.py

```diff
@@ -236,48 +249,79 @@ def run(ceph_cluster: Ceph, **kwargs) -> int:
+                    LOG.info("Initiating IO before scale down")
                     for initiator in ha.clients:
                         io_tasks.append(executor.submit(initiator.start_fio))
                     time.sleep(20)  # time sleep for IO to Kick-in
 
-                    ha.scale_down(gateway_nodes)
+                    ha.scale_down(gateway_nodes_to_be_deployed)
 
                 # Scale up
                 if lb_config.get("scale_up"):
                     scaleup_nodes = lb_config["scale_up"]
                     gateway_nodes = config["gw_nodes"]
+                    LOG.info(f"Started scaling up {scaleup_nodes}")
 
-                    # Prepare FIO Execution
-                    gw_node = get_node_by_id(ceph_cluster, config["gw_nodes"][-1])
-                    gateways = NVMeGateway(gw_node, ha.mtls)
-                    namespaces = ha.fetch_namespaces(gateways)
+                    # Prepare FIO execution for existing namespaces
+                    old_namespaces = ha.fetch_namespaces(ha.gateways[0])
                     ha.prepare_io_execution(initiators)
 
-                    # Check for targets at clients
-                    ha.compare_client_namespace([i["uuid"] for i in namespaces])
-
-                    # Start IO Execution
-                    LOG.info("Initiating IO before scale up")
+                    # Start IO Execution into already existing namespaces/nodes
+                    LOG.info("Initiating IO before scale up ")
                     for initiator in ha.clients:
                         io_tasks.append(executor.submit(initiator.start_fio))
                     time.sleep(20)  # time sleep for IO to Kick-in
 
                     # Perform scale-up of new nodes
-                    if not all(
-                        [node in set(config["gw_nodes"]) for node in scaleup_nodes]
-                    ):
+                    if not all([node in set(gateway_nodes) for node in scaleup_nodes]):
+
                         # Perform scale up
-                        namespaces = ha.scale_up(scaleup_nodes)
-                        for scaleup_node in scaleup_nodes:
-                            gw_node = get_node_by_id(ceph_cluster, scaleup_node)
-                            ha.gateways.append(NVMeGateway(gw_node))
-                            for subsys_args in config["subsystems"]:
-                                configure_listeners(ha, [scaleup_node], subsys_args)
+                        old_namespaces = parse_namespaces(config, old_namespaces)
+                        ha.scale_up(scaleup_nodes, gw_nodes, old_namespaces)
+
+                        # Validate IO for old namespaces
+                        ha.validate_scaleup(scaleup_nodes, old_namespaces)
+
+                        # Add listeners and namespaces to newly added GWs
+                        for subsys_args in config["subsystems"]:
+                            LOG.info(f"Adding listeners for {scaleup_nodes}")
```

Comment:
@HaruChebrolu It is better to add all listeners first and then move to new namespace addition in interest of time and IO getting stuck . Though they are new GWs, it is always advised to add all listeners before adding namespaces

Reply:
The order below would be more beneficial for scale up new Gws

1. Run IO on old namespaces 
2. ceph orch apply new GW nodes
3. Add listeners on all subsystems for alll new GW nodes immediately
4. Then add namespaces
5. Disconnect, discover and connect again from initiators

Currently it is 2,1,4-3,5

Reply:
Will update the loop to add listeners to all subsystems and then move to adding namespaces.
And currently I running IO on old namespaces and then I am applying new GW nodes, please let me know if I am missing something.

Reply:
Thanks @HaruChebrolu 
I got confused with https://github.com/red-hat-storage/cephci/blob/a16e3a89f636bd905fc9900e35973122b9e48a54/tests/nvmeof/test_ceph_nvmeof_loadbalancing.py#L280-L283

i think you are validating IO for old namespaces after orch apply new nodes? If its case then its fine. thanks
Another suggestionis you can validate IO for old namespaces as well as new namespaces together after listener add and namespace add. We can do this way too

Reply:
Done